receiver/nginxreceiver/internal/scraper/accesslog/grok.go

```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package accesslog
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/trivago/grok"
+	"go.uber.org/zap"
+)
+
+var (
+	formatVariables = map[string]string{
+		"$remote_addr":              "%{IPORHOST:remote_addr}",
+		"$remote_user":              "%{USERNAME:remote_user}",
+		"$time_local":               `%{HTTPDATE:time_local}`,
+		"$status":                   "%{INT:status}",
+		"$body_bytes_sent":          "%{NUMBER:body_bytes_sent}",
+		"$http_referer":             "%{DATA:http_referer}",
+		"$http_user_agent":          "%{DATA:http_user_agent}",
+		"$http_x_forwarded_for":     "%{DATA:http_x_forwarded_for}",
+		"$bytes_sent":               "%{NUMBER:bytes_sent}",
+		"$gzip_ratio":               "%{DATA:gzip_ratio}",
+		"$server_protocol":          "%{DATA:server_protocol}",
+		"$request_length":           "%{INT:request_length}",
+		"$request_time":             "%{DATA:request_time}",
+		"\"$request\"":              "\"%{DATA:request}\"",
+		"$request ":                 "%{DATA:request} ",
+		"$upstream_connect_time":    "%{DATA:upstream_connect_time}",
+		"$upstream_header_time":     "%{DATA:upstream_header_time}",
+		"$upstream_response_time":   "%{DATA:upstream_response_time}",
+		"$upstream_response_length": "%{DATA:upstream_response_length}",
+		"$upstream_status":          "%{DATA:upstream_status}",
+		"$upstream_cache_status":    "%{DATA:upstream_cache_status}",
+		"[":                         "\\[",
+		"]":                         "\\]",
+	}
+
+	// Pattern to match all the variables that are mentioned in the access log format
+	logVarRegex = regexp.MustCompile(`\$([a-zA-Z]+[_[a-zA-Z]+]*)`)
+)
+
+func newGrok(logFormat string, logger *zap.SugaredLogger) (*grok.CompiledGrok, error) {
```

Comment:
Ideally we would want this code to be exported in the Agent and we would then import it here. However, I think it needs our `grok` wrapper would need a little bit more work cuz it's not very robust right now. The moment the `log_format` is slightly off, `grok` doesn't know what to do with the log entries and just returns no data. It also does so _silently_, meaning the end user will not get any feedback that that their `log_format` is off should `grok` fail to parse log entries.